openedx-unsupported/ecommerce
# Redeem coupon and enroll user in course

Context:

We recently created a "browse and request" feature that allows enterprise learners to request access to courses. We only assign a coupon code as part of the approval process, but we want to enroll the learners directly into the course. Ecommerce will consume coupon-request-approval events sent by enterprise-access (via ecommerce-plugin-events which will be installed in a separate pr), and redeem the code on behalf of the learner. 

I extracted the logic from `CouponRedeemView` so that coupons can be redeemed outside the context of a request.
It mimics the regular code redemption flow, one that an enterprise learner would go through the learner portal.


To test:
1. Assign a coupon to an enterprise learner via the admin portal
2. Shell into ecommerce via `make dev.shell.ecommerce` and run `'./manage.py shell` to get into the django shell
3. In the django shell, run `import default_code_redeemer from ecommerce.coupons.api` && `default_code_redeemer.redeem_coupon_code(coupon_code, course_id, lms_user_id)`
4. You should see the code redeemed successfully in the console and that the user has been enrolled into the course **without** a data-sharing consent record created.

# ⛔️ DEPRECATION WARNING

**This repository is deprecated and in maintainence-only operation while we work on a replacement, please see [this announcement](https://discuss.openedx.org/t/deprecation-removal-ecommerce-service-depr-22/6839) for more information.**

Although we have stopped integrating new contributions, we always appreciate security disclosures and patches sent to [security@edx.org](mailto:security@edx.org)



## Anyone internally merging to this repository is expected to [release and monitor their changes](https://openedx.atlassian.net/wiki/spaces/RS/pages/1835106870/How+to+contribute+to+our+repositories); if you are not able to do this DO NOT MERGE, please coordinate with someone who can to ensure that the changes are released.

## Description

Describe what this pull request changes, and why these changes were made. How will these changes affect other people, installations of edx, etc.?
Please include links to any relevant ADRs, design artifacts, and decision documents. Make sure to document the rationale behind significant changes in the repo, per [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be
linked here.

Useful information to include:
- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author", "Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these changes.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

Please provide detailed step-by-step instructions for testing this change; how did YOU test this change?

## Other information

Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, OpenEdx vs. edx.org differences, development vs. production environment differences, security, or accessibility.
